ofrak_patch_maker/ofrak_patch_maker/patch_maker.py

```diff
@@ -229,43 +236,66 @@ def make_bom(
         :raises PatchMakerException: if user inputs are invalid.
         :return: an immutable object containing section info
         """
+        set_start_method("spawn", force=True)
+
         if self._platform_includes:
             header_dirs.extend(self._platform_includes)
         self._validate_bom_input(name, source_list, object_list, header_dirs)
         object_map = {}
         for o_file in object_list:
-            assembled_object = self.prepare_object(o_file)
+            assembled_object = self.prepare_object(self, o_file)
             object_map.update({o_file: assembled_object})
 
         out_dir = os.path.join(self.build_dir, name + "_bom_files")
         os.mkdir(out_dir)
 
         for c_file in list(filter(lambda x: x.endswith(".c"), source_list)):
             assembled_object_path = self._toolchain.compile(c_file, header_dirs, out_dir=out_dir)
-            assembled_object = self.prepare_object(assembled_object_path)
+            assembled_object = self.prepare_object(self, assembled_object_path)
             object_map.update({c_file: assembled_object})
 
-        for asm_file in list(filter(lambda x: x.endswith(".as") or x.endswith(".S"), source_list)):
-            original_asm_file = asm_file
-            if asm_file.endswith(".S"):
-                asm_file = self._toolchain.preprocess(asm_file, header_dirs, out_dir=out_dir)
-            assembled_object_path = self._toolchain.assemble(asm_file, header_dirs, out_dir=out_dir)
-            assembled_object = self.prepare_object(assembled_object_path)
-            object_map.update({original_asm_file: assembled_object})
+        # Call global function which runs parallel assembly
+        asm_files = list(filter(lambda x: x.endswith(".as") or x.endswith(".S"), source_list))
+        args = zip(
+            asm_files,
+            itertools.repeat(header_dirs),
+            itertools.repeat(out_dir),
+            itertools.repeat(self._toolchain),
+            itertools.repeat(self.prepare_object),
+            itertools.repeat(self),
+        )
+        workers = math.ceil(cpu_count())
+        with get_context("spawn").Pool(processes=workers) as pool:
+            result = pool.starmap(
+                _assemble_source_files, args, chunksize=math.ceil(len(asm_files) / workers)
+            )
+        for r in result:
+            object_map.update(r)
```

Comment:
Might as well do this for .c and .o as well, though I'm curious what the debug log of the toolchain executions looks like now. Probably not cute.

Reply:
Can you measure the subprocess spawn overhead and evaluate how many .o files we'd need to try to parse before parallelism helps us?

Reply:
I recommend doing some reading on Python benchmarking best practices, especially as they may pertain to measuring other programs' execution time.

You may want to write your own short Python script with your own event loop to isolate from other OFRAK runtime variability...